### PR TITLE
TB 408: Create an initial UI for the Re-processor pipeline

### PR DESCRIPTION
**Pull request by ItsMurumba**
- Restructure of the mediator into monorepo
- Introduced Microfrontend UI for the Reprocessor
- Adds Dockerfile for containerization of the UI
- Updates Github ACtions for build process

### CODE REVIEW

File: apps/reprocessor-mediator-microfrontend/src/components/StepOne.tsx [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  Card,
+  CardContent,
+  Select,
+  Checkbox,
+  ListItemText,
+  FormControl,
+  InputLabel,
+} from "@mui/material";
+
+const availableResources = [
```
Comment (brett-onions) [inline apps/reprocessor-mediator-microfrontend/src/components/StepOne.tsx:17]: can you add MedicationStatement, ServiceRequest, MedicationDispense, QuestionnaireResponse, Procedure, DiagnosticReport

Comment (ItsMurumba) [reply apps/reprocessor-mediator-microfrontend/src/components/StepOne.tsx:17]: Sure.

File: .gitignore [hunk 2]
```diff
@@ -1 +1,2 @@
 node_modules
+dist
```
Comment (bradsawadye) [inline .gitignore]:
```suggestion
dist

```
Add new line to end of file

File: apps/reprocess-mediator/.prettierignore [hunk 3]
```diff
@@ -1,3 +1,3 @@
 # Ignore artifacts:
 build
-coverage
+coverage
```
Comment (bradsawadye) [inline apps/reprocess-mediator/.prettierignore]:
```suggestion
coverage

```

File: apps/reprocess-mediator/config/config.mjs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  process.env.OPENHIM_MEDIATOR_URL || 'https://localhost:8080'
+export const OPENHIM_TRANSACTION_URL =
+  process.env.OPENHIM_TRANSACTION_URL || 'http://localhost:5001'
+export const OPENHIM_USERNAME =
+  process.env.OPENHIM_USERNAME || 'root@openhim.org'
+export const OPENHIM_PASSWORD = process.env.OPENHIM_PASSWORD || 'instant101'
+export const OPENHIM_CLIENT_CUSTOM_TOKEN =
+  process.env.OPENHIM_CLIENT_CUSTOM_TOKEN || 'test'
+
+export const TRUST_SELF_SIGNED = process.env.TRUST_SELF_SIGNED || true
```
Comment (bradsawadye) [inline apps/reprocess-mediator/config/config.mjs]: Environment variables are of string type. The expression should evaluate to a boolean. The value will only be a boolean (true) when the env variable is not specified. If the env variable is specified to true then you have the value "true" which is not equivalent to boolean true. 
```suggestion
export const TRUST_SELF_SIGNED = process.env.TRUST_SELF_SIGNED === 'false' ? false : true
```

File: apps/reprocess-mediator/config/config.mjs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+export const KAFKA_CONCURRENCY = stringToInt(process.env.KAFKA_CONCURRENCY, 2)
+export const KAFKA_REPROCESSING_TOPIC =
+  process.env.KAFKA_REPROCESSING_TOPIC || 'reprocess'
+export const KAFKA_2XX_TOPIC = process.env.KAFKA_2XX_TOPIC || '2xx'
+export const KAFKA_URL = process.env.KAFKA_URL || 'kafka-01:9092'
+
+// MongoDB
+
+export const MONGODB_CONNECTION_STRING = process.env.MONGODB_CONNECTION_STRING || 'mongodb://localhost/openhim';
+export const MONGODB_DIRECT_CONNECTION = process.env.MONGODB_DIRECT_CONNECTION || false;
```
Comment (bradsawadye) [inline apps/reprocess-mediator/config/config.mjs]: should evaluate to a boolean

File: apps/reprocess-mediator/mediatorConfig.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "urn": "urn:mediator:reprocess-mediator",
-  "version": "v1.0.0",
+  "version": "1.4.1",
```
Comment (bradsawadye) [inline apps/reprocess-mediator/mediatorConfig.json]: not matching version in the package.json

File: apps/reprocess-mediator/mediatorConfig.json [hunk 7]
```diff
@@ -80,11 +102,35 @@
           "type": "bool"
         },
         {
-          "param": "episodeOfCare",
+          "param": "espisodeOfCare",
```
Comment (bradsawadye) [inline apps/reprocess-mediator/mediatorConfig.json]:
```suggestion
          "param": "episodeOfCare",
```

File: apps/reprocess-mediator/routes/index.mjs [hunk 8]
```diff
@@ -6,11 +6,17 @@ import {buildReturnObject} from './utils.mjs'
...
+import cors from 'cors';
 
 const routes = express.Router()
+routes.use(cors());
 
+routes.get('/reprocess/mongo', getHandler);
+routes.post('/reprocess/mongo', postHandler);
 routes.post('/reprocess', reprocessRoute)
 
+
```
Comment (bradsawadye) [inline apps/reprocess-mediator/routes/index.mjs]:
```suggestion
```
Consistent spacing

File: apps/reprocess-mediator/routes/mongo.mjs [hunk 9]
```diff
@@ -0,0 +1,133 @@
+import mongoose from 'mongoose'
+import { sendDataToKafka } from './kafka.mjs';
+import { KAFKA_2XX_TOPIC, MONGODB_CONNECTION_STRING, MONGODB_DIRECT_CONNECTION } from '../config/config.mjs';
+import logger from '../logger.mjs';
+
+
```
Comment (bradsawadye) [inline apps/reprocess-mediator/routes/mongo.mjs]: consistent spacing

File: apps/reprocess-mediator/routes/mongo.mjs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  let payload
+  if(method === 'POST'){
+     logger.info("Processing POST requests");
+     payload = await formatPostRequests(query, resources);
+
+  }else {
+    logger.info("Processing DELETE requests");
+    payload = await formateDeleteRequests(query);
+  }
```
Comment (bradsawadye) [inline apps/reprocess-mediator/routes/mongo.mjs]: format this

File: apps/reprocess-mediator/routes/mongo.mjs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+          payload,
+          reject,
+          resolve,
+          KAFKA_2XX_TOPIC)
+      });
+    } catch (error) {
+      logger.error(`Failed to push to Kafka. Error Message: ${error.message}`);
+      failedCounter++;
+    }
+
```
Comment (bradsawadye) [inline apps/reprocess-mediator/routes/mongo.mjs]:
```suggestion
```

File: apps/reprocess-mediator/routes/mongodbRoute.mjs [hunk 12]
```diff
@@ -0,0 +1,77 @@
+import { queryOpenhimTransactions, pushMessagesToKafka } from './mongo.mjs';
+
+export async function getHandler(req, res) {
+  const { query } = req
+
+  const validatedQuery = validateReprocessingOptions(query);
+
+  if(validatedQuery.error){
+    return res.status(400).json({error: validatedQuery.error})
+
```
Comment (bradsawadye) [inline apps/reprocess-mediator/routes/mongodbRoute.mjs]:
```suggestion
```

File: apps/reprocess-mediator/routes/mongodbRoute.mjs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  // Get the results from MongoDB
+  const { numberOfFhirResources, numberOfTransactions } = await queryOpenhimTransactions(validatedQuery);
+
```
Comment (bradsawadye) [inline apps/reprocess-mediator/routes/mongodbRoute.mjs:14]: inconsistent spacing

File: apps/reprocess-mediator/routes/mongodbRoute.mjs [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+  }
+  else if(validatedBody.method === 'DELETE'){
+    const { 
+      numberOfTransactions,
+      requests
+     } = await queryOpenhimTransactions(validatedBody);
+
+      const failedCounter = await pushMessagesToKafka(requests);
+
+     return res.json({
```
Comment (bradsawadye) [inline apps/reprocess-mediator/routes/mongodbRoute.mjs]: Please format this

File: apps/reprocess-mediator/routes/mongodbRoute.mjs [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+  
+    const resources = reprocessingOptions?.resources;
+  
+    return {
+      fromDate: reprocessingOptions.reprocessFromDate,
+      toDate: reprocessingOptions.reprocessToDate,
+      method: reprocessingOptions.method || null,
+      resources,
+    };
+  }
```
Comment (bradsawadye) [inline apps/reprocess-mediator/routes/mongodbRoute.mjs]: end of line and please format this

File: apps/reprocessor-mediator-microfrontend/.env.local [hunk 16]
```diff
@@ -0,0 +1 @@
+REPROCESSOR_API_BASE_URL = http://localhost:3000/
```
Comment (bradsawadye) [inline apps/reprocessor-mediator-microfrontend/.env.local]:
```suggestion
REPROCESSOR_API_BASE_URL = http://localhost:3000/

```

File: apps/reprocessor-mediator-microfrontend/.prettierignore [hunk 17]
```diff
@@ -0,0 +1,8 @@
+.gitignore
+.prettierignore
+yarn.lock
+yarn-error.log
+package-lock.json
+dist
+coverage
+pnpm-lock.yaml
```
Comment (bradsawadye) [inline apps/reprocessor-mediator-microfrontend/.prettierignore]:
```suggestion
pnpm-lock.yaml

```

**Review by bradsawadye — CHANGES_REQUESTED** [review]
Looks good. Left some minor comments mostly on formatting the code

**Review by bradsawadye — APPROVED** [review]

File: apps/reprocessor-mediator-microfrontend/config/nginx.conf [hunk 18]
```diff
@@ -0,0 +1,9 @@
+server {
+    listen 80;
+    root /usr/share/nginx/html;
+    index index.html;
+
+    location / {
+        try_files $uri $uri/ /index.html;
+    }
+}
```
Comment (bradsawadye) [inline apps/reprocessor-mediator-microfrontend/config/nginx.conf]: new line at the end

**Review by brett-onions — APPROVED** [review]

**Pull request merged**